File: SGPhotoPickerController/SGPhotoPickerController.m
//
//  SGPhotoPickerController.m
//  Whip Around
//
//  Created by Igor Anany on 31/03/16.
//  Copyright © 2016 7 glyphs Ltd. All rights reserved.
//

#import "SGPhotoPickerController.h"

@interface SGPhotoPickerController () <UIImagePickerControllerDelegate, UINavigationControllerDelegate>

@property (nonatomic, assign) BOOL canGetFromCamera;
@property (nonatomic, strong) UIViewController * controller;

- (void)takeControll;
- (void)resignControll;
- (void)selectPhoto;
- (void)takePhoto;

@end

@implementation SGPhotoPickerController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (instancetype)initWithViewController:(UIViewController *)controller {
    self = [super init];
    if (self) {
        self.controller = controller;
        self.canGetFromCamera = [UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera];
        self.cameraEnabled = YES;
        self.libraryEnabled = YES;
    }
    return self;
}

- (void)takeControll {
    [self.controller addChildViewController:self];
}

- (void)resignControll {
    [self removeFromParentViewController];
}

- (void)pickPhoto {
    [self takeControll];
    if (self.libraryEnabled && self.cameraEnabled && self.canGetFromCamera) {
        UIAlertController *actionSheet = [UIAlertController alertControllerWithTitle:@"Add photo" message:nil preferredStyle:UIAlertControllerStyleActionSheet];
        
        [actionSheet addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction *action) {
            [self resignControll];
            [actionSheet dismissViewControllerAnimated:YES completion:^{
            }];
        }]];
        
        [actionSheet addAction:[UIAlertAction actionWithTitle:@"Photo Library" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
            [self selectPhoto];
            [actionSheet dismissViewControllerAnimated:YES completion:^{
                
            }];
        }]];
        
        [actionSheet addAction:[UIAlertAction actionWithTitle:@"Camera" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
            [self takePhoto];
            [actionSheet dismissViewControllerAnimated:YES completion:^{
                
            }];
        }]];
        
        [self.controller presentViewController:actionSheet animated:YES completion:nil];
    }
    else {
        if (self.cameraEnabled && !self.canGetFromCamera) {
            NSLog(@"Camera source is not available");
            if (self.libraryEnabled) {
                [self selectPhoto];
            }
        }
        else if (self.cameraEnabled && self.canGetFromCamera) {
            [self takePhoto];
        }
        else if (self.libraryEnabled) {
            [self selectPhoto];
        }
    }
}

- (void)selectPhoto {
    UIImagePickerController *picker = [[UIImagePickerController alloc] init];
    picker.delegate = self;
    picker.allowsEditing = YES;
    picker.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
    
    [self.controller presentViewController:picker animated:YES completion:NULL];
}

- (void)takePhoto {
    UIImagePickerController *picker = [[UIImagePickerController alloc] init];
    picker.delegate = self;
    picker.allowsEditing = YES;
    picker.sourceType = UIImagePickerControllerSourceTypeCamera;

    [self.controller presentViewController:picker animated:YES completion:NULL];
}

#pragma mark - UIImagePickerControllerDelegate

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary *)info {
    if (self.delegate) {
        UIImage *chosenImage = info[UIImagePickerControllerEditedImage];
        [self.delegate photoPickerControllerDidPickImage:self image:chosenImage];
    }
    [picker dismissViewControllerAnimated:YES completion:^{
        [self resignControll];
    }];
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker {
    [picker dismissViewControllerAnimated:YES completion:^{
        [self resignControll];
    }];
}
@end

File: SGPhotoPickerController/SGPhotoPickerController.swift
//
//  SGPhotoPickerController.swift
//  SGPhotoPickerControllerExample
//
//  Created by Igor Anany on 7/09/16.
//  Copyright © 2016 7 glyphs Ltd. All rights reserved.
//

import UIKit

protocol SGPhotoPickerControllerDelegate: class {
    func photoPickerController(_ photoPickerController: SGPhotoPickerController, didPickImage image: UIImage)
}

class SGPhotoPickerController: UIViewController {
    
    // Private properties
    
    fileprivate var canGetFromCamera: Bool
    fileprivate var controller: UIViewController
    
    // Public properties
    
    weak var delegate: SGPhotoPickerControllerDelegate?
    var object: AnyObject?
    var libraryEnabled = true
    var cameraEnabled = true
    
    // Private methods
    
    fileprivate func takeControll() {
        self.controller.addChildViewController(self)
    }
    
    fileprivate func resignControll() {
        self.removeFromParentViewController()
    }
    
    fileprivate func openPicker(_ type: UIImagePickerControllerSourceType) {
        let picker = UIImagePickerController()
        picker.delegate = self
        picker.allowsEditing = true
        picker.sourceType = type
        self.controller.present(picker, animated: true, completion: nil)
    }
    
    // Public methods
    
    init(controller: UIViewController) {
        self.controller = controller
        self.canGetFromCamera = UIImagePickerController.isSourceTypeAvailable(.camera)
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func pickPhoto() {
        self.takeControll()
        if self.libraryEnabled && self.cameraEnabled && self.canGetFromCamera {
            let actionSheet = UIAlertController(title: "Add photo", message: nil, preferredStyle: .actionSheet)
            actionSheet.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: { (action) in
                self.resignControll()
                actionSheet.dismiss(animated: true, completion: nil)
            }))
            actionSheet.addAction(UIAlertAction(title: "Photo Library", style: .default, handler: { (action) in
                self.openPicker(.photoLibrary)
                actionSheet.dismiss(animated: true, completion: nil)
            }))
            actionSheet.addAction(UIAlertAction(title: "Camera", style: .default, handler: { (action) in
                self.openPicker(.camera)
                actionSheet.dismiss(animated: true, completion: nil)
            }))
            self.controller.present(actionSheet, animated: true, completion: nil)
        }
        else {
            if self.cameraEnabled && !self.canGetFromCamera {
                print("Camera source is not available")
                if self.libraryEnabled {
                    self.openPicker(.photoLibrary)
                }
            }
            else if self.cameraEnabled && self.canGetFromCamera {
                self.openPicker(.camera)
            }
            else if self.libraryEnabled {
                self.openPicker(.photoLibrary)
            }
        }
    }
}

extension SGPhotoPickerController: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
        let image = info[UIImagePickerControllerEditedImage] as! UIImage
        self.delegate?.photoPickerController(self, didPickImage: image)
        picker.dismiss(animated: true) {
            self.resignControll()
        }
    }
    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        picker.dismiss(animated: true) { 
            self.resignControll()
        }
    }
}
